### DISCUSSION TITLE
Mapping GeoJSON resulting in "TypeError: Cannot convert undefined or null to object"

### DISCUSSION THREAD

**Discussion by PieterBoeren** [nartc/mapper#388] [Q&A] [answered]
I'm trying to create a mapping profile to map from GeoJSON to Proj4GeoJSONFeature, but the mapping fails at the point where the mapper is being created with this error:

![image](https://user-images.githubusercontent.com/43775697/147093230-3a613537-7990-4e2a-9b45-4d462c44d502.png)

I'm using the Transformer plugin to decorate my classes, but the mapping is just failing with the above error. I also tried using the decorator, but that is failing since Proj4GeoJSONFeature is a type (and not an interface).

DTO:
```Typescript
class TestDto {
    public id!: number;
    public someGeometry!: GeoJSON.GeoJSON;
}
```
ViewModel:
```Typescript
import { Proj4GeoJSONFeature } from "proj4leaflet";

class TestViewModel {
    public id!: number;
    public someGeometry!: Proj4GeoJSONFeature;
}
```

Profile:

```Typescript
import { Converter, convertUsing, MappingProfile } from "@automapper/core";
import { Proj4GeoJSONFeature } from "proj4leaflet";
import { TestDto } from "./models.dto";
import { TestViewModel } from "./models.vm";

const testProfile: MappingProfile = (mapper) => {
  mapper.createMap(TestDto, TestViewModel)
  .forMember(
    (destination) => destination.someGeometry,
    //mapFrom((source) => source.someGeometry)
    convertUsing(geoJsonConverter, (source) => source.someGeometry)
  );
};

const geoJsonConverter: Converter<GeoJSON.GeoJSON, Proj4GeoJSONFeature> = {
  convert(source) {
    debugger;
    return { type: "Feature", geometry: { type: "Point", coordinates: [125.6, 10.1] }, properties: [] };
  },
};
```
Example code:

```Typescript
// Example data from https://geojson.org/
const geoJson: GeoJSON.GeoJSON = {
  "type": "Feature",
  "geometry": {
    "type": "Point",
    "coordinates": [125.6, 10.1]
  },
  "properties": {
    "name": "Dinagat Islands"
  }
};

mapper.map({id:1, someGeometry: geoJson}, TestViewModel, TestDto);
```

One thing to notice is that the `debugger` statement in the converter isn't even called, it's failing even before the data is actually being converted.

I have created a reproduction repository here: https://github.com/PieterBoeren/cra-automapper.

I might need some more advanced configuration that the transformer plugin does support, but I really can't figure out what that configuration should be. Any clues?

**Comment by nartc** [comment]
I'll take a look as soon as I have some free time.

**Comment by PieterBoeren** [comment]
In noticed that marking the "geo related" types with the AutoMap decorator with just about anything (Object/Number/Boolean/etc) works. Doesn't feel correct, but at least I'm getting into the custom type converter (example shows Object, but Boolean works just as fine).

```TypeScript
import { AutoMap } from "@automapper/classes";
import { Proj4GeoJSONFeature } from "proj4leaflet";

class TestViewModel {
    public id!: number;
    @AutoMap({ typeFn: () => Object })
    public someGeometry!: Proj4GeoJSONFeature;
}

export { TestViewModel };
```

**Reply by micalevisk** [reply]
If `Proj4GeoJSONFeature` is a type definition, not a concrete class, then you must use `typenFn`, otherwise there will be no valid metadata to AutoMapper look at into due to TypeScript limitation.

**Reply by PieterBoeren** [reply]
It is indeed not a concrete class, but a Type. The thing that seems strange to me is that putting *anything* in the AutoMap just works, but leaving it out doesn't work...

**Reply by nartc** [reply]
Sorry for the confusion. I think I can handle that better in automapper core. However, if `someGeometry` is **ALWAYS** going to be configured with `forMember()` aka **MANUAL** configuration then `@AutoMap()` isn't needed.

**Reply by PieterBoeren** [reply]
Now I'm getting more confused 😉

Are you saying that when I use a `forMember()`, the decorator isn't necessary? Since my code was doing exactly that, and failing, it only works with both the decorator and the `forMember()`.

**Reply by nartc** [reply]
Interesting. That definitely sounds like a bug. Can I have a reproduce?

**Reply by PieterBoeren** [reply]
It is in the repository in my start post.

**Reply by micalevisk** [reply]
> Are you saying that when I use a forMember(), the decorator isn't necessary?

yeah. I missed the `forMember` for `someGeometry`, that's why I said you needed the `@AutoMap()` before, my bad.

**Reply by nartc** [reply]
@PieterBoeren Released 7.1.1 with a quick fix. Though my statement is still correct, that is if a field is **MANUALLY** configured, you don't need `@AutoMap()`. So your model and vm should be:

```ts
class TestDto {
    @AutoMap()
    public id!: number;
    public someGeometry!: GeoJSON.GeoJSON;
}

class TestViewModel {
    @AutoMap()
    public id!: number;
    public someGeometry!: Proj4GeoJSONFeature;
}
```

since `id` is being `auto-mapped`, we need `@AutoMap()` for it. Another thing is you'd need to turn on: 
```json
"experimentalDecorators": true,
"emitDecoratorMetadata": true,
```
in `tsconfig.json`

![image](https://user-images.githubusercontent.com/25516557/147296657-192dae67-77e1-43b9-998d-bec75a62fe26.png)

**Reply by PieterBoeren** [reply]
Thanks, the quick fix did indeed solve my issue (Since I'm using the transformer plugin, I do not need to use the decorator on the `id` column. The fantastic plugin takes care of that)!

I indeed missed the extra tsconfig setting, I looked it up in the documentation, and in my honest opinion, the installation document could be improved:

The [installation instructions](https://automapperts.netlify.app/docs/getting-started/installation) do not mention the required tsconfig changes (if using the decorator), but the [introduction](https://automapperts.netlify.app/docs/getting-started/introduce-to-automapper) page does, and has the installation repeated, but not complete (missing the reflect-metadata).
Wouldn't it be much more clear if all the required installation is actually on the installation page and the other pages just refer to that?

I'm more than happy to create a PR if you agree (next week somewhere).

**Reply by nartc** [reply]
Oh nice, I didn't know you're using the transformer (didn't really notice xD). If you're using the transformer then I know exactly what the problem was. I need to have the transformer handle type/interface better then.

About docs, yes, please do if you have the time. I'm not a good tech writer at all 😅

**Reply by PieterBoeren** [reply]
Unfortunately, I encountered a new issue, created a bug report this time: https://github.com/nartc/mapper/issues/389

**Comment by PieterBoeren** [comment, marked as answer]
Since I can only select my own answer above as answer to the entire discussion (which contains code which isn't longer necessary): the issue has been fixed in version 7.1.1. Thanks @nartc!
